src/connector/src/sink/google_pubsub.rs

```diff
@@ -0,0 +1,279 @@
+// limitations under the License.
+
+use std::collections::HashMap;
+use std::usize;
+
+use anyhow::{anyhow, Context};
+use google_cloud_googleapis::pubsub::v1::PubsubMessage;
+use google_cloud_pubsub::client::{Client, ClientConfig};
+use google_cloud_pubsub::publisher::Publisher;
+use risingwave_common::array::StreamChunk;
+use risingwave_common::catalog::Schema;
+use risingwave_common::session_config::sink_decouple::SinkDecouple;
+use serde_derive::Deserialize;
+use serde_with::serde_as;
+use tonic::Status;
+use with_options::WithOptions;
+
+use super::catalog::desc::SinkDesc;
+use super::catalog::{SinkEncode, SinkFormatDesc};
+use super::formatter::SinkFormatterImpl;
+use super::log_store::DeliveryFutureManagerAddFuture;
+use super::writer::{
+    AsyncTruncateLogSinkerOf, AsyncTruncateSinkWriter, AsyncTruncateSinkWriterExt, FormattedSink,
+};
+use super::{DummySinkCommitCoordinator, Result, Sink, SinkError, SinkParam, SinkWriterParam};
+use crate::dispatch_sink_formatter_str_key_impl;
+
+pub const PUBSUB_SINK: &str = "google_pubsub";
+
+#[serde_as]
+#[derive(Clone, Debug, Deserialize, WithOptions)]
+pub struct GooglePubSubConfig {
+    /// The Google Pub/Sub Project ID
+    #[serde(rename = "pubsub.project_id")]
+    pub project_id: String,
+
+    /// Specifies the Pub/Sub topic to publish messages
+    #[serde(rename = "pubsub.topic")]
+    pub topic: String,
+
+    /// The Google Pub/Sub endpoint URL
+    #[serde(rename = "pubsub.endpoint")]
+    pub endpoint: String,
+
+    /// use the connector with a pubsub emulator
+    /// <https://cloud.google.com/pubsub/docs/emulator>
+    #[serde(rename = "pubsub.emulator_host")]
+    pub emulator_host: Option<String>,
+
+    /// A JSON string containing the service account credentials for authorization,
+    /// see the [service-account](https://developers.google.com/workspace/guides/create-credentials#create_credentials_for_a_service_account) credentials guide.
+    /// The provided account credential must have the
+    /// `pubsub.publisher` [role](https://cloud.google.com/pubsub/docs/access-control#roles)
+    #[serde(rename = "pubsub.credentials")]
+    pub credentials: Option<String>,
+
+    // accept "append-only"
+    pub r#type: String,
+}
+
+impl GooglePubSubConfig {
+    fn from_hashmap(values: HashMap<String, String>) -> Result<Self> {
+        serde_json::from_value::<GooglePubSubConfig>(
+            serde_json::to_value(values).expect("impossible"),
+        )
+        .map_err(|e| SinkError::Config(anyhow!(e)))
+    }
+
+    pub(crate) fn initialize_env(&self) {
+        tracing::debug!("setting pubsub environment variables");
+        if let Some(emulator_host) = &self.emulator_host {
+            std::env::set_var("PUBSUB_EMULATOR_HOST", emulator_host);
+        }
+        if let Some(credentials) = &self.credentials {
+            std::env::set_var("GOOGLE_APPLICATION_CREDENTIALS_JSON", credentials);
+        }
+    }
```

Comment:
I think you can also try set the emulator host and credentials via [`ClientConfig`](https://docs.rs/google-cloud-pubsub/latest/google_cloud_pubsub/client/struct.ClientConfig.html) instead of env var, to eliminate the possibility of env var conflict.

Reply:
+1, it is quite strange that we set env variable for a sink job. Can you share more on why we need this ENV?

Reply:
In order to set the environment in `ClientConfig`, we need to introduce another direct dependency (now indirect): https://docs.rs/google-cloud-gax/0.17.0/google_cloud_gax/conn/enum.Environment.html. `google-cloud-pubsub` dont re-export it (still) :|
If it's ok, I'll set it up via this.

Reply:
No problem.